src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,188 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.next = next;
+            this.value = value;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(tail, value,null);
+        if (this.isEmpty()) {
+            head = newNode;
+            tail = head;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bound: " + index);
+        }
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            if (this.isEmpty()) {
+                tail = head = newNode;
+                size++;
```

Comment:
The `size++` operation is repeated in both the `if` and `else` branches for the case when `index == 0`. It would be more efficient to move this operation outside of the conditional block.